File: src/main/java/io/protobase/f7/formulas/statistical/COUNTBLANK.java
package io.protobase.f7.formulas.statistical;

import io.protobase.f7.formulas.AbstractFormula;
import io.protobase.f7.formulas.FormulaName;
import io.protobase.f7.models.GridColumnRowKey;
import io.protobase.f7.utils.Mappers;

import java.util.Objects;
import java.util.function.BiFunction;
import java.util.function.Function;
import java.util.stream.Collectors;
import java.util.stream.Stream;

/**
 * The count of number values in a set of data.
 */
public class COUNTBLANK extends AbstractFormula {
  public static FormulaName NAME = FormulaName.COUNTBLANK;
  public static COUNTBLANK SELF = new COUNTBLANK();

  public COUNTBLANK() {
    super();
  }

  public COUNTBLANK(Function<Object, Object> lookup, BiFunction<GridColumnRowKey, Object, Object> collateralLookup) {
    super(lookup, collateralLookup);
  }

  @Override
  public Object internal(GridColumnRowKey origin, Object... values) {
    checkAtLeastLength(values.length, 1, NAME);
    return ((double) Stream.of(values)
        .map(lookup)
        .flatMap(Mappers::toFlatStream)
        .filter(Objects::isNull)
        .collect(Collectors.toList()).size());
  }
}

File: src/main/js/formulas/statistical/COUNTBLANK.ts
import { isNull } from "../../utils/Other";
import { SheetColumnRowKey } from "../../models/common/SheetColumnRowKey";
import { Complex } from "../../models/common/Types";
import { Counter } from "../../utils/Counter";
import { Mappers } from "../../utils/Mappers";
import { Reducers } from "../../utils/Reducers";
import { AbstractFormula } from "../AbstractFormula";
import { FormulaName } from "../FormulaName";

export class COUNTBLANK extends AbstractFormula {
  static SELF: COUNTBLANK = new COUNTBLANK();
  NAME = FormulaName.COUNTBLANK;

  internal(origin: SheetColumnRowKey, ...values: Array<Complex>) {
    AbstractFormula.checkAtLeastLength(values.length, 1, this.NAME);
    const counter = new Counter<Complex>();
    values
      .map(this.lookup)
      .map(Mappers.flattenGridsToArrays) // 1, 2, [null, 1, "0", "Nope.", null]
      .map(Mappers.ensureAllAreArrays) // [1], [2], [null, 1, "0", null]
      .reduce(Reducers.join) // [1, 2, null, 1, "0", null]
      .filter(isNull) // [null, null]
      .map((value) => counter.count(value)); // [null, null]
    return counter.getCount();
  }
}
